File: neet_code_all/src/array_hash/majority_element.rs
use std::collections::HashMap;

struct Solution {}
impl Solution {
    /// AC
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let n = nums.len() as i32;
        let mut map: HashMap<i32, i32> = HashMap::new();

        for v in &nums {
            *map.entry(*v).or_default() += 1;
        }

        let mut majo_idx = nums[0];
        for (idx, num) in map {
            if num > n / 2 {
                majo_idx = idx;
                break;
            }
        }

        majo_idx
    }

    // AC
    pub fn majority_element_2(nums: Vec<i32>) -> i32 {
        let n = nums.len() as i32;
        let mut map: HashMap<i32, i32> = HashMap::new();

        let mut majo_idx = nums[0];
        for v in &nums {
            *map.entry(*v).or_default() += 1;
            if let Some(majo_num) = map.get(v) {
                if *majo_num > n / 2 {
                    majo_idx = *v;
                    break;
                }
            }
        }

        majo_idx
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    // time: O(N), space: O(1)
    // Boyer-Moore のアルゴリズム
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let (mut res, mut count) = (0, 0);

        for n in nums {
            if count == 0 {
                res = n;
            }
            count += if n == res { 1 } else { -1 };
        }

        res
    }

    // time: O(N), space: O(N)
    pub fn majority_element_2(nums: Vec<i32>) -> i32 {
        let mut count: HashMap<i32, i32> = HashMap::new();
        let (mut res, mut max_count) = (0, 0);

        for num in nums {
            *count.entry(num).or_insert(0) += 1;
            res = if *count.get(&num).unwrap() > max_count {
                num
            } else {
                res
            };

            max_count = i32::max(*count.get(&num).unwrap(), max_count);
        }

        res
    }

    // time: O(NlogN), space: O(1)
    pub fn majority_element_3(mut nums: Vec<i32>) -> i32 {
        nums.sort();
        nums[nums.len() / 2]
    }
}

fn main() {
    let case_1 = vec![3, 2, 3];
    // => 3
    let case_2 = vec![2, 2, 1, 1, 1, 2, 2];
    // => 2

    println!("case_1: {}", Solution::majority_element(case_1.clone()));
    println!("case_1: {}", Solution::majority_element_2(case_1.clone()));
    println!("case_2: {}", Solution::majority_element(case_2.clone()));
    println!("case_2: {}", Solution::majority_element_2(case_2.clone()));

    println!("case_1: {}", SolutionAns::majority_element(case_1.clone()));
    println!(
        "case_1: {}",
        SolutionAns::majority_element_2(case_1.clone())
    );
    println!(
        "case_1: {}",
        SolutionAns::majority_element_3(case_1.clone())
    );
    println!("case_2: {}", SolutionAns::majority_element(case_2.clone()));
    println!(
        "case_2: {}",
        SolutionAns::majority_element_2(case_2.clone())
    );
    println!(
        "case_2: {}",
        SolutionAns::majority_element_3(case_2.clone())
    );
}

File: neet_code_all_cpp/array_hash/majority_element.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
    public:
        // AC
        int majorityElement(std::vector<int>& nums) {
            auto n = nums.size();
            std::unordered_map<int,int> map;

            for(auto& v: nums) {
                ++map[v];
            }

            auto majo_idx = nums[0];
            for(auto& [idx, num]: map) {
                if(num > n / 2) {
                    majo_idx = idx;
                    break;
                }
            }

            return majo_idx;
        }

        // AC
        int majorityElement2(std::vector<int>& nums) {
            auto n = nums.size();
            std::unordered_map<int,int> map;

            auto majo_idx = nums[0];
            for(auto& v: nums) {
                ++map[v];
                if(map.find(v) != map.end()) {
                    if(map[v] > n / 2) {
                        majo_idx = v;
                        break;
                    }
                }
            }

            return majo_idx;
        }
};

// Rustの模範解答より
class SolutionAnsRust {
    public:
        // time: O(N), space: O(1)
        // Boyer-Moore algorithm
        // AC
        int majorityElement(std::vector<int>& nums) {
            auto [res, count] = std::pair(0, 0);

            for(auto& n : nums) {
                if(count == 0) {
                    res = n;
                }

                if(n == res) {
                    ++count;
                } else {
                    --count;
                }
            }

            return res;
        }

        // AC
        // time: O(N), space: O(n)
        int majorityElement2(std::vector<int>& nums) {
            std::unordered_map<int, int> map;
            auto [res, max_count] = std::pair(0, 0);

            for(auto& num: nums) {
                ++map[num];

                if(map.find(num) != map.end()) {
                    if(map[num] > max_count) {
                        res = num;
                    }
                }

                max_count = std::max(map[num], max_count);
            }

            return res;
        }

        // AC
        // time: O(NlogN), space: O(1)
        int majorityElement3(std::vector<int> nums) {
            std::sort(nums.begin(), nums.end());
            return nums[nums.size() / 2];
        }
};

// 模範解答
class SolutionAns {
    public:
        int majorityElement(std::vector<int>& nums) {
            auto [count, res] = std::pair(0, 0);

            for(auto& num: nums) {
                if(!count) {
                    res = num;
                }

                count += (num == res) ? 1: -1;
            }

            return res;
        }
};

int main(void) {
    std::vector<int> case_1 = {3, 2, 3};
    // => 3
    std::vector<int> case_2 = {2,2,1,1,1,2,2};
    // => 2

    Solution s_1;
    std::cout << s_1.majorityElement(case_1) << std::endl;
    std::cout << s_1.majorityElement(case_2) << std::endl;

    std::cout << s_1.majorityElement2(case_1) << std::endl;
    std::cout << s_1.majorityElement2(case_2) << std::endl;

    SolutionAnsRust s_ans_rs;
    std::cout << s_ans_rs.majorityElement(case_1) << std::endl;
    std::cout << s_ans_rs.majorityElement(case_2) << std::endl;

    std::cout << s_ans_rs.majorityElement2(case_1) << std::endl;
    std::cout << s_ans_rs.majorityElement2(case_2) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.majorityElement(case_1) << std::endl;
    std::cout << s_ans.majorityElement(case_2) << std::endl;
} 
